gh: expose throttler parameters as options

petr-muller opened the pull request:
Tools using the client may want to configure the behavior of their
throttling (these values are currently usually hardcoded). GH client
parameters are configured in a shared structure, but different tools
that use the client may want to specify different defaults depending on
their use case.

Add a new method that allows to build a flag parser using specified
defaults. I added a method using the functional options pattern so it is
extensible and quite easily usable from callsites.

The change makes all clients constructed from GH options call
`Throttle`, but the defaults are zeroes, which result in unthrottled
clients. The existing users of `Throttle` will not be broken, they will
just call `Throttle` twice, once with defaults, once with their values,
so their behavior should not change. The existing users can be ported to
defaults by removing their hardcoded `Throttle` calls and providing
custom defaults for options. I included such change for the
`needs-rebase` plugin and will port other callsites in test-infra if the
PR is accepted.

alvaroaleman commented on prow/flagutil/github.go:
```diff
@@ -47,25 +47,62 @@ type GitHubOptions struct {
+}
+
+// AddFlagsWithCustomDefaults injects GitHub options into the given FlagSet,
+func (o *GitHubOptions) AddFlagsWithCustomDefaults(fs *flag.FlagSet, defaulters ...CustomGithubDefaulter) {
```
How do we deal with tide which constructs two clients with different defaults for their throttler settings? And also I would have thought we have more components with pre-existing flags for this

petr-muller replied:
> How do we deal with tide which constructs two clients with different defaults for their throttler settings?

The change should at least not break it, `Throttle` is smart enough and it can handle multiple subsequent `Throttle` calls, so Tide should just continue working same as it did before. The problematic part is probably that the components that do something smart like this will have confusing `--help` output and values passed in this new param would not be respected. Would that be a blocker?

> And also I would have thought we have more components with pre-existing flags for this

We probably do, but I haven't invested the effort in porting them all. I will do the legwork if the general approach is accepted.

alvaroaleman replied:
> The problematic part is probably that the components that do something smart like this will have confusing --help output and values passed in this new param would not be respected. Would that be a blocker?

Can we maybe just add an option to not add the new throttler flags if requested and use that?

> We probably do, but I haven't invested the effort in porting them all. I will do the legwork if the general approach is accepted.

sounds reasonable

petr-muller replied:
One way how to make Tide conform is to infer its partial throttlers from the totals given by the params. That way Tide will first create the clients trottled to the specified values (right now the clients are unthrottled) and then subsequent `Throttle` calls cal further restrict the individual clients.

petr-muller replied:
> Can we maybe just add an option to not add the new throttler flags if requested and use that?

Yep. It does not fit into the current pattern but it should not be hard to bend it for that purpose.

alvaroaleman commented on prow/flagutil/github.go:
```diff
@@ -47,25 +47,62 @@ type GitHubOptions struct {
+	fs.StringVar(&o.TokenPath, "github-token-path", defaults.TokenPath, "Path to the file containing the GitHub OAuth secret.")
+	fs.StringVar(&o.AppID, "github-app-id", defaults.AppID, "ID of the GitHub app. If set, requires --github-app-private-key-path to be set and --github-token-path to be unset.")
+	fs.StringVar(&o.AppPrivateKeyPath, "github-app-private-key-path", defaults.AppPrivateKeyPath, "Path to the private key of the github app. If set, requires --github-app-id to bet set and --github-token-path to be unset")
+	fs.IntVar(&o.ThrottleHourlyTokens, "github-hourly-tokens", defaults.ThrottleHourlyTokens, "If set to a value larger than zero, enable client-side throttling to limit hourly token consumption. If set, --github-allowed-burst must be positive too.")
```
I was kind of expecting the flag adding to be conditional

petr-muller replied:
Why? I can make it so, but is there a good reason to make them so?

alvaroaleman replied:
Yes, something like tide will have multiple flags for this and one of it just doesn't do anything which is confusing at best

petr-muller replied:
Maybe I misunderstood, but I thought you suggested the https://github.com/kubernetes/test-infra/pull/21891#discussion_r617678748 to allow opting out.

alvaroaleman replied:
I wrote:
> Can we maybe just add an option to not add the new throttler flags if requested and use that?

Adding a bool option to the flagutil to not add these two flags if explicitly requested is what I meant by that

petr-muller commented:
I ported all users of `Throttle()`. Tide is the only thing that wants to do its own thing and suppress the common options. The only other leftover is `label_sync` which does not use the common option group at all, so it would need to be ported to it entirely and I do not intend to do that in the PR.

1. Need to fix a bunch of tests, mostly related to existing same-purpose options in existing tools. 
2. Need to fix the old + new option resolution, what I wrote originally for needs-rebase is trash

petr-muller commented:
/wip

alvaroaleman commented on prow/flagutil/github.go:
```diff
@@ -47,25 +47,87 @@ type GitHubOptions struct {
+// effectively disallowing external users to parametrize default throttling
+// behavior. This is useful mostly when a program creates multiple GH clients
+// with different behavior.
+func DenyThrottlerOptions() FlagParameter {
```
```suggestion
func DisableThrottlerOptions() FlagParameter {
```

petr-muller replied:
#drunkcoding but I do not drink anymore so /shrug

petr-muller replied:
Fixed everywhere.

alvaroaleman commented on prow/cmd/checkconfig/main.go:
```diff
@@ -271,7 +276,6 @@ func validate(o options) error {
 		if err != nil {
 			return fmt.Errorf("error loading GitHub client: %w", err)
 		}
-		githubClient.Throttle(3000, 100) // 300 hourly tokens, bursts of 100
```
I think this has a zero too much :D

petr-muller replied:
Heh. It's this way since 7 Aug 2019 when I made that mistake in https://github.com/kubernetes/test-infra/pull/13627 so maybe let's keep it if nobody complained until now?

alvaroaleman commented on prow/cmd/branchprotector/protect.go:
```diff
@@ -47,9 +47,12 @@ type options struct {
-	tokenBurst         int
-	github             flagutil.GitHubOptions
+
+	// TODO(petr-muller): Remove after August 2021, replaced by github.ThrottleHourlyTokens
```
Maybe add a deprecation warning to ANNOUNCEMENTS.md?

petr-muller replied:
Sounds good!

petr-muller commented on prow/flagutil/github.go:
```diff
@@ -93,6 +155,18 @@ func (o *GitHubOptions) Validate(bool) error {
 
+	if (o.ThrottleHourlyTokens > 0) != (o.ThrottleAllowBurst > 0) {
+		if o.ThrottleHourlyTokens == 0 {
+			// Tolerate `--github-hourly-tokens=0` alone to disable throttling
```
Well spotted, this is a little ugly.

Initially I had it like you say: either both are positive, or both are zero. I changed it to the current state to not break users. The existing tools that 1) throttled by default 2) had custom throttling settings that allowed to set hourly tokens to zero to disable throttling allowed a call like this, without setting the burst option to zero too:

```
$ peribolos --hourly-tokens=0 ...
```

In this PR, these legacy options are translated to the new common ones, so existing usage like above would be broken if we enforced that both params are zero.

petr-muller commented on prow/external-plugins/needs-rebase/main.go:
```diff
@@ -73,9 +85,11 @@ func gatherOptions() options {
 	fs.StringVar(&o.webhookSecretFile, "hmac-secret-file", "/etc/webhook/hmac", "Path to the file containing the GitHub HMAC secret.")
 	fs.StringVar(&o.logLevel, "log-level", "debug", fmt.Sprintf("Log level is one of %v.", logrus.AllLevels))
-	fs.IntVar(&o.hourlyTokens, "hourly-tokens", 360, "The number of hourly tokens need-rebase may use")
+	fs.IntVar(&o.hourlyTokens, "hourly-tokens", defaultHourlyTokens, "The number of hourly tokens need-rebase may use. DEPRECATED: use --github-allowed-burst")
```
We do, on L72

petr-muller commented on prow/flagutil/github.go:
```diff
@@ -47,25 +47,87 @@ type GitHubOptions struct {
+
+	if !params.denyThrottlerOptions {
+		fs.IntVar(&o.ThrottleHourlyTokens, "github-hourly-tokens", defaults.ThrottleHourlyTokens, "If set to a value larger than zero, enable client-side throttling to limit hourly token consumption. If set, --github-allowed-burst must be positive too.")
+		fs.IntVar(&o.ThrottleAllowBurst, "github-allowed-burst", defaults.ThrottleAllowBurst, "Size of token consumption bursts. If set, --github-hourly-tokens must be positive too and set to a higher or equal number.")
```
In the current implementation that default will either be zero (and `ThrottleHourlyTokens` will be too) when the program does not customize the defaults, or `ThrottleAllowBurst` will be explicitly set to something via `ThrottlerDefaults`.

petr-muller commented:
Thanks for the review!  Addressed all feedback by either code changes (resolved convos) or I responded with explaining why some parts are done the way they are.

alvaroaleman reviewed (approved):
thanks! I am going to squash this as none of the commits can be reverted individually
/label tide/merge-method-squash
